### PR TITLE
"Compatibly" reimplement disk_usage API and logic

### PR DESCRIPTION
**Pull request by e2**
## Summary

Implement `ArubaPath#blocks` by guessing when it's not available for a given filesystem.

NOTE: If the block size cannot be determined, a default of 4096 is used. (I'm aware this is different from the default of `Aruba.config.physical_block_size`, which has a default of 512 - I believe it should be 4096).

## Details

The formula is basically asking "how many blocks are needed to at least fit the file's size?".

A spec was added to test various edge values.

## Motivation and Context

If `File::Stat.blocks` is nil, guessing is a better option than nothing.

Also, "disk usage" based on "blocks reported by stat()" is unreliable because of filesystem metadata and things like compression/deduplication (e.g. BTRFS).

So since disk usage is a "guess" at best anyway, I implemented this by calculating the blocks based on the file size. Primitive filesystems (e.g FAT) are likely more conservative with blocks, so it seems reasonable.

And there are network filesystem where we'll "never know" anyway.

Futher more, network filesystems won't even tell you what the `blksize` is, so that has to be "guessed" too.

I assumed: 4096 bytes. This seems reasonable because:

1. Only exotic filesystems and ones with custom settings have different values.
2. It's a sane balance between many small files and losing too much disk space.
3. Most Windows filesystems (where it's the usually case where values aren't available) usually defaults to 4k anyway
4. People likely aren't going to be using Aruba for calculating disk space on humongous filesystems
5. Most large-blocked local filesystems usually will likely be advanced enough to provide the info available anyway
6. Systems with smaller block sizes will be "overestimated", which is still a good thing. This doesn't matter much, because either the usage will exceed the device size (or be close), so people will notice ... or ... for lots of large files filling up the media, the total "margin of error" will be small.
7. Even mission-critical systems aren't likely going to rely on disk usage, which is to error-prone in theory alone.

So I see no downside. 

## How Has This Been Tested?

Specs added and they pass (locally). File::Stat is mocked, so whatever my own filesystem reports is irrelevant. 

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

(The "guessing" could be unexpected, but assuming disk usage measurement can be accurate is ... incorrect anyway).

## Checklist:

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] Documentation has been updated

### CODE REVIEW

**Comment by e2** [comment]
Ok, turns out Ruby's documentation is misleading. (Ruby uses info provided by `stat()` directly).

blocks is **always** the number of 512 byte blocks allocated. It's always 512 bytes.

`blksize` is the "recommended size of IO", not the filesystem's allocation unit. It's "recommended" which means it's close to what's called "clusters" in the Windows world. So this value is useless if `blocks` are given, and it's probably absent anyway if `blocks` are `nil`.

And so, the `Aruba.physical_block_size` is simply misleading. It's not an allocation unit, so it isn't relevant for calculating.

In fact, it should be renamed to "fs_allocation_unit_size" and it should be 4096 by default. This is because we're calculating "disk usage". Almost no filesystem has 512 as the "block size", but almost every stat() implementation relies on `blocks` being the amount of allocated 512-sized blocks.

What you really want to know is: "what's the minimum sized block for an empty or 1-byte file?".

On Linux, this is: `blocks` * 512. That's why a 1-byte file usually has "8", so it's actually taking up 8*512=4096, which is the "allocation unit" of 4096.

This means `ArubaPath.blocks` should always return `File::Stat.blocks * 8 / Aruba.fs_allocation_unit_size`. Except that `File::Stat.blksize` is a more reliable way of discovering the `fs_allocaiton_unit_size` for any given **specific** file.

(I'm not sure about network systems).

I think the whole disk_usage API has to be redesigned. It's based on `File::Stat` values, which is the most misleading thing ever.

**Comment by e2** [comment]
The commits and commits should be self-explanatory.

The biggest paradigm shift is burying the concept of `blocks` from `File::Stat`.

Instead, the number of `blocks` should be immediately converted into a "filesystem unit".

That's the difference: confusing "blocks" vs understandable "units".

Explanation:

On a tiny file, Stat.blocks returns 8, which is 8*512 bytes, which is 4096.

A file can't take up 1,2,3,4,5,6 or 7 "blocks" on my filesystem. Only 8, 16, 32, etc.
Instead, files can take up 1,2,3,... "units" of 4kB.

The changes are 100% compatible (except for the extra method to show total bytes allocated - which superseeds the method returning "blocks").

Anyway, the rest is in the PR.

**Comment by dg-ratiodata** [comment]
Reading all your comments through, I think adding this functionality really was a mistake of mine:

* It's unreliable
* It works different on different combinations of OS AND FS

I really ask myself, why not deprecating and removing it. It doesn't look like it's worth to keep. It's far to complicated and hard to understand. WDYT? Sorry, you really did a good job by improving this functionality.

**Comment by e2** [comment]
It's part of the API and you can't remove it until 1.x. 

So might as well have it merged to fix the issue (on AppVeyor at least) and consider dropping it from 1.x.

> It's unreliable

In terms of code, probably not. In terms of the idea of being able to estimate disk usage - it's as good as it get. 

There's `du` in the coreutils (and on Cygwin/Ming) but using it would be a lot trickier with no major upside. I'm not sure about use cases, but if there is a use case for checking disk usage, it's good enough.

> It works different on different combinations of OS AND FS

Yes and not - if the data isn't provided, estimates are quite good. The worst case is if you have:
1. Many 1-bytes files everywhere ...
2. ... which are truncated huge files ...
3. ... on a log-based FS (probably).

The issues are not much worse than with file permissions/chmod.

So by dropping this, you should also consider dropping chmod. It's unlikely to be useful across platforms, except an API like:
- "make_readonly", "make_writable"
- "make_executable", "make not executable" (POSIX)

Everything else (groups, etc.) means multi-user, and you probably can't switch users without admin access, which implies running Aruba as root, which implies and exotic setup that isn't portable.

If this is dropped due to complexity, same thing should happen with chmod. Even if they're both moved to aruba-contrib or something.

File: features/api/filesystem/disk_usage.feature [hunk 1]
```diff
@@ -1,14 +1,45 @@
...
+  the `stat()` system call will report 8 `blocks` of 512 bytes, which is
+  exactly 4096 bytes. So ignore `blocks` values - unless you multiply them by
+  512 (almost always).
+
+  Currently Aruba allows you to set what's called the `physical_block_size`,
+  which is a bit misleading - it's the "512" value. So if you somehow have a
+  "filesystem unit size" of 8192 (instead of a typical 4KB), set the
+  `physical_block_size` to 1024 (yes, divide by 8: 8192/8 = 1024).
+
+  Ideally, in the future Aruba could provide `Aruba.config.fs_allocation_unit`,
```
Comment (dg-ratiodata) [inline features/api/filesystem/disk_usage.feature]: Let's don't add this here. This might confuse users.

File: lib/aruba/aruba_path.rb [hunk 2]
```diff
@@ -120,14 +120,47 @@ def [](index)
...
+    TYPICAL_FS_UNIT = 4096
+    TYPICAL_DEV_BSIZE = 512 # Google DEV_BSIZE for more info
+
     # Report count of blocks allocated on disk
     #
     # This reports the amount of blocks which are allocated by the path.
     #
     # @return [Integer]
     #   The count of blocks on disk
+    # TODO: deprecate this method
```
Comment (dg-ratiodata) [inline lib/aruba/aruba_path.rb]: Add `@deprecated` here.

File: lib/aruba/aruba_path.rb [hunk 3]
```diff
@@ -120,14 +120,47 @@ def [](index)
...
+    end
+
+    # Report minimum disk space used
+    #
+    # This reports the minimum bytes allocated by the path.
+    #
+    # @return [Integer]
+    #   Total bytes allocate
+    # TODO: deprecate this method
+    # TODO: this is recommended over the above "blocks"
```
Comment (dg-ratiodata) [inline lib/aruba/aruba_path.rb]: Add `@deprecated` here.

**Comment by dg-ratiodata** [comment]
> It's part of the API and you can't remove it until 1.x. 

Sure. Deprecate and then remove.

> So by dropping this, you should also consider dropping chmod. 

Mmh... Personally I use this a lot. From my understanding it generally works, but on Windows only for a specific set of permissions, e.g. `644` etc. I would "solve" that via documentation. Let's have a separate issue for API-cleanup.

Please address my comments and we're good to go.

**Comment by e2** [comment]
I reworked the comments a bit. 

I also removed the deprecation from the new method I introduced. (I can't judge if the feature is useful or not).

Technically, blocks are a more reliable way to determine the used space on Linux. On Windows, the filesystem's "cluster size" (filesystem allocation unit) is the only thing you can really use. So it would be trading accuracy for portability. Not to mention it's less confusing on the surface.

(E.g. A small EXT2 partition typically has 1kB blocks).

If you want to keep it, I'd make Aruba's configuration value `nil` by default, and use `File::Stat.blksize` whenever possible. This would make things correct on directory trees spanning multiple filesystems with different allocation unit sizes.

Anyway, this should be a separate gem.

**Comment by e2** [comment]
Can this be merged?

File: lib/aruba/api/filesystem.rb [hunk 4]
```diff
@@ -372,10 +372,25 @@ def with_file_content(file, &block)
       #
       # @result [FileSize]
       #   Bytes on disk
+
+      TYPICAL_FS_UNIT = 4096 # Very typical, except for giant or embedded filesystems
```
Comment (maxmeyer) [inline lib/aruba/api/filesystem.rb]: Are those constants used somewhere else? If not, let's make them just variables. I find that easier to maintain.

File: lib/aruba/aruba_path.rb [hunk 5]
```diff
@@ -126,8 +126,61 @@ def [](index)
...
-      File::Stat.new(to_s).blocks
+      min_bytes_used = minimum_disk_space_used
+      min_bytes_used / Aruba.config.physical_block_size
+    end
+
+    # TODO: Aruba.config.physical_block_size could be allowed to be nil
+    # (So the unit size can be autodetected)
+
+    TYPICAL_FS_UNIT = 4096
+    TYPICAL_DEV_BSIZE = 512 # Google DEV_BSIZE for more info
```
Comment (maxmeyer) [inline lib/aruba/aruba_path.rb]: Please make them variables if not use somewhere else.

**Comment by maxmeyer** [comment]
Just replace the constants by variables and I'm going to merge this.

**Comment by maxmeyer** [comment]
Ok. Did it myself to reduce the "backlog".

**Comment by e2** [comment]
Thanks.
